test/integration/puppetlabs/puppetserver/error_handling_int_test.clj

```diff
@@ -57,3 +61,36 @@
                      (:body response)))
               (is (re-matches #"text/plain; charset=.*"
                               (get-in response [:headers "content-type"]))))))))))
+
+(deftest ^:integration test-HTTP-404
+  (testing "When an improperly-formatted request is made (i.e., in the pre-Puppet 4 format),
+            a 404 error is logged containing information on upgrading the agent."
+    (bootstrap/with-puppetserver-running
+      app
+      {:certificate-authority {:certificate-status {:authorization-required false}}}
+      (with-test-logging
+        (let [jruby-service (tk-app/get-service app :JRubyPuppetService)
+              puppet-server-version-regex (re-pattern (str "Puppet Server version: "
+                                                           (version-check/get-version-string "puppet-server")))
+              puppet-version-regex (re-pattern (str "Puppet version: " (:puppet-version (config/get-puppet-config jruby-service))))
+              puppet-api-regex (re-pattern (str "Supported /puppet API versions: " master/puppet-API-versions))
+              puppet-ca-api-regex (re-pattern (str "Supported /puppet-ca API versions: " master/puppet-ca-API-versions))]
+          (testing "Puppet 3 Master URLs are no longer valid"
+            (let [response (http-client/get
+                             "https://localhost:8140/production/catalog/localhost"
+                             bootstrap/request-options)]
+              (is (= 404 (:status response)))
+              (is (re-find puppet-server-version-regex (:body response)))
+              (is (re-find puppet-version-regex (:body response)))
+              (is (re-find puppet-api-regex (:body response)))
+              (is (re-find puppet-ca-api-regex (:body response)))))
+
+          (testing "Puppet 3 CA URLs are no longer valid"
```

Comment:
These are great, but we need one more test where a request is made that has never been valid.  For example, when I request `/foo/bar/baz/this/has/never/been/vaid`, I get a response that does _not_ inform the user to upgrade the agent.